### PR TITLE
osd: fix osd detection logic for old lsblk

### LINKED ISSUE #10665: Trying to create more OSDs than asked for
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

In a host based cluster with OSDs located on LVMs, after the cluster was successfully built, a reconciliation run is trying to add even more OSDs, and failing:

```
2022-08-02 09:49:35.891930 E | op-osd: failed to provision OSD(s) on node fra1pro-k8s-worker-6.k8s.fra1pro.financial.com. &{OSDs:[] Status:failed PvcBackedOSD:false Message:failed to configure devices: failed to initialize osd: failed to run ceph-volume raw command. Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new c3c6e3a2-a838-4910-a6b2-6075ce841c7e
Running command: /usr/bin/ceph-authtool --gen-print-key
Running command: /usr/bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-6
Running command: /usr/bin/chown -R ceph:ceph /dev/dm-5
Running command: /usr/bin/ln -s /dev/dm-5 /var/lib/ceph/osd/ceph-6/block
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-6/activate.monmap
 stderr: got monmap epoch 5
Running command: /usr/bin/ceph-authtool /var/lib/ceph/osd/ceph-6/keyring --create-keyring --name osd.6 --add-key AQAs8+hi7vErNxAAEXzVZx/Kx0S5Uq+6uv4+og==
 stdout: creating /var/lib/ceph/osd/ceph-6/keyring
added entity osd.6 auth(key=AQAs8+hi7vErNxAAEXzVZx/Kx0S5Uq+6uv4+og==)
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-6/keyring
Running command: /usr/bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-6/
Running command: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 6 --monmap /var/lib/ceph/osd/ceph-6/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-6/ --osd-uuid c3c6e3a2-a838-4910-a6b2-6075ce841c7e --setuser ceph --setgroup ceph
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 bluestore(/var/lib/ceph/osd/ceph-6/) _open_db_and_around failed to load os-type: (2) No such file or directory
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 bluestore(/var/lib/ceph/osd/ceph-6/) mkfs fsck found fatal error: (2) No such file or directory
 stderr: 2022-08-02T09:49:34.249+0000 7f3c95eb1200 -1 OSD::mkfs: ObjectStore::mkfs failed with error (2) No such file or directory
 stderr: 2022-08-02T09:49:34.250+0000 7f3c95eb1200 -1  ** ERROR: error creating empty object store in /var/lib/ceph/osd/ceph-6/: (2) No such file or directory
--> Was unable to complete a new OSD, will rollback changes
Running command: /usr/bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring osd purge-new osd.6 --yes-i-really-mean-it
 stderr: purged osd.6
Traceback (most recent call last):
  File "/usr/sbin/ceph-volume", line 11, in <module>
    load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 40, in __init__
    self.main(self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
    return f(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 152, in main
    terminal.dispatch(self.mapper, subcommand_args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/main.py", line 32, in main
    terminal.dispatch(self.mapper, self.argv)
  File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
    instance.main()
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 169, in main
    self.safe_prepare(self.args)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 91, in safe_prepare
    self.prepare()
  File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
    return func(*a, **kw)
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 134, in prepare
    tmpfs,
  File "/usr/lib/python3.6/site-packages/ceph_volume/devices/raw/prepare.py", line 68, in prepare_bluestore
    db=db
  File "/usr/lib/python3.6/site-packages/ceph_volume/util/prepare.py", line 481, in osd_mkfs_bluestore
    raise RuntimeError('Command failed with exit code %s: %s' % (returncode, ' '.join(command)))
RuntimeError: Command failed with exit code 250: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 6 --monmap /var/lib/ceph/osd/ceph-6/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-6/ --osd-uuid c3c6e3a2-a838-4910-a6b2-6075ce841c7e --setuser ceph --setgroup ceph: exit status 1}
2022-08-02 09:49:35.891958 I | op-k8sutil: format and nodeName longer than 63 chars, nodeName fra1pro-k8s-worker-6.k8s.fra1pro.financial.com will be 18a40a276db7618c5ffd6fd2d8ca0b8c
```

However, there are already 5 valid OSDs:

```
[rook@rook-ceph-tools-69c54d54bc-rz6zs /]$ ceph osd tree
ID   CLASS  WEIGHT    TYPE NAME                                                     STATUS  REWEIGHT  PRI-AFF
 -1         15.00000  root default                                                                           
 -8          3.00000      rack pa18                                                                          
 -7          3.00000          host fra1pro-k8s-worker-6-k8s-fra1pro-financial-com                            
  1    hdd   3.00000              osd.1                                                 up   1.00000  1.00000
 -4          6.00000      rack pa19                                                                          
-19          3.00000          host fra1pro-k8s-worker-8-k8s-fra1pro-financial-com                            
  3    hdd   3.00000              osd.3                                                 up   1.00000  1.00000
 -3          3.00000          host fra1pro-k8s-worker-9-k8s-fra1pro-financial-com                            
  4    hdd   3.00000              osd.4                                                 up   1.00000  1.00000
-12          3.00000      rack pa20                                                                          
-11          3.00000          host fra1pro-k8s-worker-10-k8s-fra1pro-financial-com                           
  0    hdd   3.00000              osd.0                                                 up   1.00000  1.00000
-16          3.00000      rack pa25                                                                          
-15          3.00000          host fra1pro-k8s-worker-12-k8s-fra1pro-financial-com                           
  2    hdd   3.00000              osd.2                                                 up   1.00000  1.00000
```

And we also didn't ask for more than 5 OSDs:

```
  storage:
    nodes:
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-6.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-8.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-9.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-10.k8s.fra1pro.financial.com
      resources: {}
    - devices:
      - name: /dev/mapper/vg_root-lv_k8s--rook--osd
      name: fra1pro-k8s-worker-12.k8s.fra1pro.financial.com
      resources: {}
    useAllDevices: false
```
(note: the helm chart's values.yaml (see attachment had `useAllNodes: false`, and I just now set this on the cluster CRD, but that was after the failed reconciliation run, and after the value wasn't added by helm)

**Expected behavior:**

Only the specified number of nodes is created.

**How to reproduce it (minimal and precise):**
Happened with two different 1.23.7 clusters, both on Rook 1.9.8.

**Environment**:
* OS (e.g. from /etc/os-release): Oracle Enterprise Linux 8.6
* Kernel (e.g. `uname -a`): `Linux fra1pro-k8s-worker-6 5.15.0-0.30.20.el8uek.x86_64 #2 SMP Sun Jul 10 12:13:30 PDT 2022 x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider or hardware configuration: HPE Proliatn DL360 Gen10
* Rook version (use `rook version` inside of a Rook Pod): rook: v1.9.8/go: go1.17.12
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 16.2.10 (45fa1a083152e41a408d15505f594ec5f1b4fe17) pacific (stable)
* Kubernetes version (use `kubectl version`): 1.23.7
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Rancher on bare metal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox]: HEALTH_OK

[operator.log](https://github.com/rook/rook/files/9241767/operator.log)
[cluster-yaml.txt](https://github.com/rook/rook/files/9241772/cluster-yaml.txt)
[ceph-cluster-values-fra1pro-yaml.txt](https://github.com/rook/rook/files/9241783/ceph-cluster-values-fra1pro-yaml.txt) (that's the helm chart's values.yaml file)

### PR DESCRIPTION
**Pull request by taki-d**
**Description of your changes:**
Old lsblk can't detect bluestore signature.
So read the head of device and verify signature
as a workaround.

**Which issue is resolved by this Pull Request:**
Resolves #10665

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [x] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/daemon/ceph/osd/daemon.go [hunk 1]
```diff
@@ -324,6 +351,17 @@ func getAvailableDevices(context *clusterd.Context, agent *OsdAgent) (*DeviceOsd
 				logger.Infof("skipping device %q because it contains a filesystem %q", device.Name, device.Filesystem)
 				continue
 			}
+		} else {
+			contain, err := containSignature(device)
+			if err != nil {
+				logger.Infof("skipping device %q. %v", device.Name, err)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon.go:357]:
```suggestion
				logger.Errorf("skipping device %q, failed to detect if it has a signature. %v", device.Name, err)
```

File: pkg/daemon/ceph/osd/daemon.go [hunk 2]
```diff
@@ -324,6 +351,17 @@ func getAvailableDevices(context *clusterd.Context, agent *OsdAgent) (*DeviceOsd
...
+				continue
+			}
+
+			if contain {
+				logger.Infof("skipping device %q because it has OSD", device.Name)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon.go:362]:
```suggestion
				logger.Infof("skipping device %q, detected an existing OSD", device.Name)
```

File: pkg/daemon/ceph/osd/daemon_test.go [hunk 3]
```diff
@@ -172,7 +173,95 @@ USEC_INITIALIZED=1128667
...
+			valid:    false,
+			hasError: true,
+		},
+	}
+
+	for _, c := range cases {
+		valid, err := containBluestoreSignature(c.device)
+
+		assert.Equal(t, c.valid, valid)
+		assert.Equal(t, c.hasError, err != nil, c.device.RealPath, err)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon_test.go:256]: How about this approach for readability (pseudocode)? As a single line it is just difficult to understand for me.
```
if c.hasError
  assert.Error(t, err)
else
  assert.NoError(t, err)
```

File: pkg/daemon/ceph/osd/daemon.go [hunk 4]
```diff
@@ -287,6 +288,32 @@ func matchDevLinks(devLinks, deviceName string) bool {
...
+	}
+	defer f.Close()
+	signature := make([]byte, len(bluestoreSignature))
+	_, err = f.Read(signature)
+	if err != nil {
+		err := fmt.Errorf("failed to read %q. %v", device.Name, err)
+		return false, err
+	}
+	if string(signature) == bluestoreSignature {
+		return true, nil
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon.go:309]: Can we detect anything besides the signature? For example, the fsid of the cluster or the osd id would be nice to log.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/daemon.go (hunk 4 above)
Comment (taki-d) [reply pkg/daemon/ceph/osd/daemon.go]: modified the PR to print OSD UUID too, thanks.

File: pkg/daemon/ceph/osd/daemon.go [hunk 5]
```diff
@@ -287,6 +291,53 @@ func matchDevLinks(devLinks, deviceName string) bool {
...
+		err := fmt.Errorf("failed to read signature from %q. %v", device.Name, err)
+		return "", err
+	}
+
+	if signature[:len(signature)-1] != bluestoreSignature {
+		return "", nil
+	}
+
+	uuid, err := reader.ReadString('\n')
+
```
Comment (travisn) [inline pkg/daemon/ceph/osd/daemon.go:330]: nit: remove blank line between checking for err != nil
```suggestion
```

**Review by travisn — APPROVED** [review]

**Review by satoru-takeuchi — APPROVED** [review]

**Pull request merged**